[#3068, #5104] Add ToC guide, enchantment changes, pack sorting (#5134)
foundryvtt/dnd5e: wiki/Table-of-Contents.md

- `chapter`: Chapters are sorted first and the entry name is displayed as a large heading. They will automatically list all of their pages below.
- `appendix`: Appendices are sorted last and also displayed as a large header. By default they do not display their individual pages, but that can be overridden using the `showPages` flag (see below).
- `special`: Special is used for journal entries that should appear as if they are pages beneath a chapter or appendix. They also hide their pages by default.
- `header`: The first page of this journal entry will be displayed inline at the top of the table of contents.

```javascript
const journalEntry = await fromUuid("...");
journalEntry.setFlag("dnd5e", "type", "chapter");
```

#### `showPages` flag

The `showPages` flag on entries controls whether the individual pages are listed beneath the entry name in the list with their own links. This defaults to `true` for entries with the `chapter` type, and `false` for entries with the `appendix` or `special` types.

```javascript
const journalEntry = await fromUuid("...");
journalEntry.setFlag("dnd5e", "showPages", false);
```

#### `position` flag

> The `position` flag is used to control where chapters and appendices appear in the list. It is usually as simple as setting the first chapter to have a `1`, the second chapter to have `2`, and so on. The position is independent between chapters and appendices, so you can have appendix A also have an position of `1`.

Reviewer: ```suggestion
The `position` flag is used to control where chapters and appendices appear in the list. It is usually as simple as setting the first chapter to have a `1`, the second chapter to have `2`, and so on. The position is independent between chapters and appendices, so you can have appendix A also have a position of `1`.
```

```javascript
const journalEntry = await fromUuid("...");
journalEntry.setFlag("dnd5e", "position", 1);
```

This flag only has effect for `chapter` and `appendix` entries.

#### `append` & `order` flags

> The `append` and `order` flags are used by `special` entries to control where they appear in the final list. The `append` flag specifies which chapter the special entry will be added to. If no `append` is provided, then a special entry will be added after all of the other chapters and appendices as a top-level entry.

Reviewer: ```suggestion
The `append` and `order` flags are used by `special` entries to control where they appear in the final list. The `append` flag specifies to which chapter the special entry will be added. If no `append` is provided, then a special entry will be added after all of the other chapters and appendices as a top-level entry.
```

Reviewer 2: English grammar does not forbid ending sentences with prepositions. The original construction was better.

![Up to date as of 4.3.0](https://img.shields.io/static/v1?label=dnd5e&message=4.3.0&color=informational)

The table of contents system allows modules to take advantage of automatic generation of table of contents view for their journal compendiums. There are a few flags that must be set by the module creator in order to take advantage of this system, but once set up the system will handle the rest and no additional code is required from the module.

![Complete table of contents page](https://raw.githubusercontent.com/foundryvtt/dnd5e/publish-wiki/wiki/images/journal/table-of-contents-complete.jpg)

### Pack Registration

> The first step to setting up the ToC is to add a flag to whichever journal compendium pack needs it. In your module's manifest (`module.json` or `world.json`) add a `flags` object to the pack definition and add a `display` flag set to `table-of-contents`:

Reviewer 2: ```suggestion
The first step to setting up the ToC is to add a flag to whichever journal compendium pack needs it. In your module's manifest (`module.json` or `world.json`) add a `flags` object to the pack definition and add a `dnd5e.display` flag set to `table-of-contents`:
```

```json
{
  "id": "dnd5e",
  "packs": [
    {
      "name": "rules",
      "label": "Rules (SRD)",
      "system": "dnd5e",
      "path": "packs/rules",
      "type": "JournalEntry",
      "private": false,

>       "flags": {
>         "display": "table-of-contents"
>       }

Reviewer 2: ```suggestion
      "flags": {
        "dnd5e": {
          "display": "table-of-contents"
        }
      }
```

- `header`: The first page of this journal entry will be displayed inline at the top of the table of contents.

```javascript
const journalEntry = await fromUuid("...");
journalEntry.setFlag("dnd5e", "type", "chapter");
```

#### `showPages` flag

The `showPages` flag on entries controls whether the individual pages are listed beneath the entry name in the list with their own links. This defaults to `true` for entries with the `chapter` type, and `false` for entries with the `appendix` or `special` types.

```javascript
const journalEntry = await fromUuid("...");
journalEntry.setFlag("dnd5e", "showPages", false);
```

#### `position` flag

The `position` flag is used to control where chapters and appendices appear in the list. It is usually as simple as setting the first chapter to have a `1`, the second chapter to have `2`, and so on. The position is independent between chapters and appendices, so you can have appendix A also have a position of `1`.

```javascript
const journalEntry = await fromUuid("...");
journalEntry.setFlag("dnd5e", "position", 1);
```

> This flag only has effect for `chapter` and `appendix` entries.

Reviewer 2: ```suggestion
This flag only affects `chapter` and `appendix` entries.
```
